tket2/src/portmatching/predicate.rs

```diff
@@ -0,0 +1,291 @@
+///
+/// See [`Predicate`] for the type of constraints that can be defined.
+pub type Constraint = pm::Constraint<HugrVariableID, Predicate>;
+
+/// Predicate for pattern matching on flat hugrs.
+///
+/// Predicates are boolean-valued functions that can be evaluated given
+/// a tuple of [`HugrVariableValue`]s.
+///
+/// When combined together with a tuple of constraint variables, a predicate
+/// defines a constraint. Given a binding from constraint variables to
+/// constraint values, a constraint can thus be evaluated by evaluating
+/// the predicate on the bound values.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, serde::Serialize, serde::Deserialize)]
+pub enum Predicate {
+    /// Unary predicate checking the node OpType
+    ///
+    /// Constraint variable must be of type [`HugrVariableID::Op`].
+    IsOpEqual(MatchOp),
+
+    /// Arity-2 Predicate checking the (unique) source of a wire
+    ///
+    /// Given a wire and a node, check that the wire is connected to the node
+    /// at the specified outgoing port.
+    ///
+    /// The first constraint variable must be of type [`HugrVariableID::Op`].
+    /// The second constraint variable must be of type
+    /// [`HugrVariableID::CopyableWire`] or [`HugrVariableID::LinearWire`].
+    IsWireSource(hugr::OutgoingPort),
+
+    /// Arity-2 Predicate checking the sink of a wire
+    ///
+    /// Given a wire and a node, check that the wire is connected to the node
+    /// at the specified incoming port.
+    ///
+    /// If the wire is a linear type, then wire sinks must be unique.
+    ///
+    /// The first constraint variable must be of type [`HugrVariableID::Op`].
+    /// The second constraint variable must be of type
+    /// [`HugrVariableID::CopyableWire`] or [`HugrVariableID::LinearWire`].
+    IsWireSink(hugr::IncomingPort),
+
+    /// 1 + `n_other` arity Predicate checking injectivity of wires (i.e. wires
+    /// are distinct).
+    ///
+    /// Given a wire and `n_other` other wires, check that the first wire is
+    /// distinct from all the other wires. This is not checking all-to-all
+    /// distinctness: this is so that the predicate is closed under AND (when
+    /// the first argument is identical).
+    ///
+    /// All `n_other` + 1 constraint variables must be of the same wire type,
+    /// either [`HugrVariableID::CopyableWire`] or [`HugrVariableID::LinearWire`].
+    IsDistinctFrom {
+        /// The number of other nodes, determining the predicate arity.
+        n_other: cmp::Reverse<usize>,
+    },
+}
+
+/// The valid "signatures" of predicates, useful for argument parsing.
+enum PredicateArguments {
+    /// Argument for a [`Predicate::IsOpEqual`] predicate.
+    Node(hugr::Node),
+    /// Arguments for a [`Predicate::IsWireSource`] or [`Predicate::IsWireSink`] predicate.
+    NodeAndWire(hugr::Node, hugr::Wire),
+    /// Arguments for a [`Predicate::IsDistinctFrom`] predicate.
+    ManyNodes(Vec<hugr::Node>),
```

Comment:
Why do you not have an option for many wires?

Reply:
I only added the variants that I needed, given the predicates and their signatures. I will rewrite this a bit to make it clearer